### ISSUE TITLE
Feature request: emoji favourites / favorites of toots

### ISSUE THREAD

**Issue by lupine** [tootsuite/mastodon#4106]
From idle musing: https://voe.social/web/statuses/112946

I can boost toots, and I can favourite toots. Favouriting is a bit like bookmarking, only less so. Visually, it assigns the ⭐ emoji to the toot.

But what if I could choose the emoji it assigned to the toot? It could be 👍 or  🍣 or anything else at all.

Toots would store the fact of the favouriting alongside the emoji, and display the emoji + counts with a query like:

```sql
SELECT emoji, count(1) FROM favourites WHERE status_id = :id GROUP BY emoji
```

To an extent, this idea comes from the GitLab.com use of 'award emoji', which are often used as an ad-hoc voting mechanism, e.g.: https://gitlab.com/gitlab-org/gitlab-ce/issues/4058

![screenshot from 2017-07-07 21-19-13](https://user-images.githubusercontent.com/115324/27975429-0a7a3d1a-635a-11e7-855f-084cd2c41740.png)

It adds a degree of richness to your interaction with the issue / toot. 

* * * *

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.

**Comment by lupine** [comment]
I'm happy to work on this feature and submit a PR if it's considered desirable. Just let me know!

**Comment by lupine** [comment]
We can migrate existing favourites by assigning the ⭐ emoji to them, of course. The `emoji` column can be `character(1)` or `varchar(1)`.

**Comment by lupine** [comment]
GitLab emoji picker like: 
![screenshot from 2017-07-07 21-34-46](https://user-images.githubusercontent.com/115324/27975913-31654fb2-635c-11e7-9eb8-d18c77d5fc88.png)

**Comment by lupine** [comment]
Mastodon's frontend could probably fit 2-3 emoji in side by side. Once a toot gets more variety than that, it'd need to be promoted to a line of its own.

**Comment by Cassolotl** [comment]
Someone suggested reaction emojis in #1178 ("like" and "save" separately), and I'd be up for that. Saves could be made private, and I could react to toots by thumbs-up-ing or hearting without having to reply or clutter my saves.

**Comment by lupine** [comment]
Thanks for the link @Cassolotl 

Emoji favourites would allow people to "save" toots by marking them with 💾 - or they could use a range of emoji to "save" toots in their own personal categorisation system.

The only remaining use case I see is the idea of allowing saves to be private. That'd need a `private` column on the `favourites` table and some soul-searching UI thought, but could reasonably be left out of a first implementation.

**Comment by marrus-sh** [comment]
Historical note: #373

**Comment by eevee** [comment]
Nit: the `emoji` column can't be `character(1)` or `varchar(1)`, because some "emoji" are actually vendor-invented ZWJ sequences of 7 or more codepoints.

Also since "emoji" is barely a real thing, maybe just let me react with any arbitrary codepoint.

Reconciling the above two paragraphs is left as an exercise.  :wink:

**Comment by fluffy-critter** [comment]
I would love to see this feature as well. How would this interact with instance-specific custom emoji?

I’d also love to see this in conjunction with UX that rolls up reactions into a bundle, Twitter-style. Also make it possible to opt out of seeing this stuff attached to your toots.

**Comment by Gargron** [comment]
Misskey federates these as Like activities with custom content (the content would be the emoji or shortcode of custom emoji, I guess). The challenge is not in the federation but the complexity and added performance overhead on retrieving the reactions. Because they are dynamic, they cannot be optimized with counter columns. It likely will remain impossible in Mastodon simply because I am not ready to make Mastodon slower and more expensive to run just for this.

**Comment by koyuawsmbrtn** [comment]
Since there is some code for reactions in the upcoming announcement feature do you plan to take this feature into consideration soon? @Gargron

**Comment by MattInternet** [comment]
Would it be sufficient to just allow "official" emojis and not instance specific ones?

**Comment by djayatillake** [comment]
> Would it be sufficient to just allow "official" emojis and not instance specific ones?

Yes I think so you only really need about 10 to allow for my emotional range. 

The LinkedIn ones are sufficient for me.

**Comment by damascene** [comment]
@koyuawsmbrtn which announcement. Is there any progress on this issue?

**Comment by koyuawsmbrtn** [comment]
> @koyuawsmbrtn which announcement. Is there any progress on this issue?

No, instance admins can do announcements and this is the only place in Mastodon where people can react to something

**Comment by alexanderadam** [comment]
Now [people are even asking to get this for Twitter](https://mobile.twitter.com/blvdmitry/status/1617302768452075521) :laughing:

**Comment by JustDalek** [comment]
Bumping support +1

It feels like we are behind on this over at mastodon compared to misskey

**Comment by Bennytek** [comment]
I would absolutely love this as well - I've got my own Misskey instance set up at the moment and I think being able to react to posts with emoji is a pretty large part of why I use it so often 😅

**Comment by JustDalek** [comment]
forreal, that's the #1 reason I hear people want to try misskey lol

**Comment by koyuawsmbrtn** [comment]
Hmm I'm thinking about having reactions Facebook-style (favs default and reaction on hover) and since Mastodon needs a little more resources such a feature can be justified now. Technology advanced quite well over the years. What's @Gargron's opinion on this topic now?

**Comment by shaedrich** [comment]
For what it's worth, on Telegram, you can configure as a group admin how many reaction emojis you allow
![grafik](https://github.com/mastodon/mastodon/assets/11225821/bdc60d19-f418-497a-b115-e81d850889e6)

Additionally, in the timeline, emoji reactions could be grouped. I don't like the idea of treating them all as a like, but categorizing them to like five basic emotions might work, though.

**Comment by UlyssesZh** [comment]
> It seems I’m in the minority here but I really don’t want this on Mastodon, it makes for a horribly busy and cluttered post UI. Stuff like this is why I use Mastodon, not other Fedi software like Misskey. 

Strictly speaking, you are free to use other UI (other clients).

**Comment by CupofDalek** [comment]
> > It seems I’m in the minority here but I really don’t want this on Mastodon, it makes for a horribly busy and cluttered post UI. Stuff like this is why I use Mastodon, not other Fedi software like Misskey.
> 
> Strictly speaking, you are free to use other UI (other clients).

Even as a supporter of this change, I think this argument works both ways. This was requested back in 2017, who knows if it will ever be added. Many features like this seem to be available on iceshrimp (misskey fork) or in catodon (iceshrimpfork)

I myself am taking this advice and not soon, but in the future hoping to switch to one of those two. Catodon's philosophy seems to align with me more, but its a newer project so I will stick to mastodon until it is out of alpha and more mature

**Comment by xarinatan** [comment]
> It seems I’m in the minority here but I really don’t want this on Mastodon, it makes for a horribly busy and cluttered post UI. Stuff like this is why I use Mastodon, not other Fedi software like Misskey.

Maybe I'm weird for suggesting this but, why not both?

You could have it toggleable to see more than just the default star, and it would work exactly the same way it already does right now (everyone on these platforms can already send these reactions, you just don't see them)


edit, I only just read this bit:
> Misskey federates these as Like activities with custom content (the content would be the emoji or shortcode of custom emoji, I guess). The challenge is not in the federation but the complexity and added performance overhead on retrieving the reactions. Because they are dynamic, they cannot be optimized with counter columns. It likely will remain impossible in Mastodon simply because I am not ready to make Mastodon slower and more expensive to run just for this.

I'd argue things have changed a bit since because other things such as GlitchSoc and others have implemented this with bounded cost: limits on reactions per user/status, top-N visible reactions by default, and batched federation via existing status cache/streaming paths rather than per-status queries.
See their patch here for more details: https://github.com/glitch-soc/mastodon/pull/2462

That should hopefully make it a lot more achievable. I can imagine performance is a key argument at Mastodon’s scale, as is avoiding clutter, but I do think there’s room to balance those concerns with what many users also want.

My personal reason for wanting this feature: I personally often run into the problem that 'favorite' has basically turned into 'I read your post and didn't hate it', and I think it would be a great addition if you want to give any bit more emotional depth to your response without actually writing a whole comment.

Could this get another look please? It's 7 years later now, maybe things have changed 🫶

**Comment by dankerthrone** [comment]
> I personally often run into the problem that 'favorite' has basically turned into 'I read your post and didn't hate it',

nail on the head and all that. 

I sometimes wonder if the core dev team even has a clear concept of what the favorite button means to users / is supposed to mean to users. Not to hate on your excellent work and all, I just don't think there's a shared common understanding of what favoriting means. And sometimes I don't want to "favorite" a toot, I just want to 😂 it, like I can in virutally every other messenger.

**Comment by dankerthrone** [comment]
I think the reactions under the orignal post in this thread speaks to peoples' desire for that function.

**Comment by quicoto** [comment]
Is there Activity Pub definition for this? How would other fediverse backends react to these reactions?

If it's supported by the spec, +1 to implement emoji reactions.

**Comment by dmoonfire** [comment]
In the spec? No. However, there is a FEP trying to establish a standard. 

https://github.com/Fedihosting-Foundation-Forks/fep/blob/main/fep/c0e0/fep-c0e0.md

**Comment by HurricaneWeasel** [comment]
I love Mastodon, but I am really missing that feature. You can't really "like" some posts: (Mental Health, Venting, …), but you don't always want to reply either. Reactions can fill a gap by allowing users to show how they feel about a post without replying to it.

Edit:
I know it's not a huge sample, but the interest in reactions is surely there.
https://mastodon.world/@HurricaneWeasel/116817291130238002

<img src="https://github.com/user-attachments/assets/9cc5006d-3b23-43c6-9811-aacfcd111e93" width="250">
